### ISSUE TITLE
The CustomQKVToContextPluginDynamic plugin uses unsatisfactory results

### ISSUE THREAD

**Issue by jacker-whh** [NVIDIA/TensorRT#2982]
## Description
I was trying to use the CustomQKVToContextPluginDynamic plugin plugin and found incorrect results. I don't know what caused it.

The input I gave is as follows:
q_tensor  [4 4 4 4]  [S B E]
k_tensor  [6 6 6 6]  [S B E]
v_tensor  [4 4 4 4]  [S B E]

The correct result returned should be [4 4 4 4]
But what the plugin returns is 
trt: [[[[[6.]]
   [[4.]]

[[6.]]

[[6.]]]]]

Is there anything wrong with my operation?

## Environment

**TensorRT Version** 8.4.11:

**NVIDIA GPU** A40:

**NVIDIA Driver Version**:

**CUDA Version**:

**CUDNN Version**:

Operating System:

Python Version (if applicable) 3.9:

Tensorflow Version (if applicable):

PyTorch Version (if applicable) 2.0:

The code is as follows：

import argparse
import ctypes
import json
import numpy as np
import os
import os.path
import re
import sys
import time
import onnx
import pycuda.autoinit
import torch
import torch.nn as nn
import torch.nn.functional as F
import common

# TensorRT
import tensorrt as trt
"""
TensorRT Initialization
"""
TRT_LOGGER = trt.Logger(trt.Logger.VERBOSE)
trt_version = [int(n) for n in trt.__version__.split('.')]
print(trt_version)

# Import necessary plugins for demoBERT
plugin_lib_name = "nvinfer_plugin.dll" if sys.platform == "win32" else "libnvinfer_plugin.so"
env_name_to_add_path = "PATH" if sys.platform == "win32" else "LD_LIBRARY_PATH"
handle = ctypes.CDLL(plugin_lib_name, mode=ctypes.RTLD_GLOBAL)
if not handle:
    raise RuntimeError("Could not load plugin library. Is `{}` on your {}?".format(
        plugin_lib_name, env_name_to_add_path))

trt.init_libnvinfer_plugins(TRT_LOGGER, "")
plg_registry = trt.get_plugin_registry()
emln_plg_creator = plg_registry.get_plugin_creator(
    "CustomEmbLayerNormPluginDynamic", "1", "")
qkv2_plg_creator = plg_registry.get_plugin_creator(
    "CustomQKVToContextPluginDynamic", "1", "")
skln_plg_creator = plg_registry.get_plugin_creator(
    "CustomSkipLayerNormPluginDynamic", "1", "")
fc_plg_creator = plg_registry.get_plugin_creator(
    "CustomFCPluginDynamic", "1", "")

attention_heads = 4
hidden_size = 4

# 定义输入张量的形状
B = 1
S = 1
E = 4
H = 2
D = 2

trt_output_model = "./fc_test_trt.fp16"

query_proj = nn.Linear(E, H*D, bias=False)
key_proj = nn.Linear(E, H*D, bias=False)
value_proj = nn.Linear(E, H*D, bias=False)
qkv_proj = nn.Linear(E, 3*H*D, bias=False)

# 生成大小为[S, B, E]的输入张量
input_tensor = torch.ones(S, B, E)
q_s = query_proj(input_tensor)  # [S,B,H*D]
# W_tmp = torch.randn(h_size, 3*h_size)

query_proj.weight.data = torch.ones(H*D, E).permute(1, 0)
key_proj.weight.data = torch.arange(0, E) + torch.zeros(H*D, E)
value_proj.weight.data = torch.ones(H*D, E).permute(1, 0)

W_tmp = torch.concatenate((query_proj.weight.permute(
    1, 0), key_proj.weight.permute(1, 0), value_proj.weight.permute(1, 0)), -1)

# print(query_proj.weight)
# print(W_tmp[:,:4]-query_proj.weight)

# 将W_tmp重塑为[E, 3, N, H]，其中N=3，H=E/3
W_tmp_reshaped = W_tmp.view(E, 3, H, D)

# [E，N，3，H]
W_tmp_transposed = W_tmp_reshaped.permute(0, 2, 1, 3)

# 将W_tmp_transposed重新整形为[E, 3*E]
W_qkv = W_tmp_transposed.reshape(E, 3*E)

# 将输入张量与权重矩阵相乘
qkv_tensor = torch.matmul(input_tensor, W_qkv).reshape(
    S, B, H, 3, D)  # [S，B，3*E，1，1] #(S, B, H, 3, D)
qkv_proj.weight.data = W_qkv.permute(1, 0)
W_aout = W_qkv.permute(1, 0).detach().numpy()
B_aout = None

with torch.no_grad():
    q_tensor, k_tensor, v_tensor = torch.split(
        qkv_tensor, split_size_or_sections=(1, 1, 1), dim=-2)

    # [B S E]
    q_tensor = q_tensor.squeeze(dim=-2).reshape(S, B, H*D).transpose(1, 0)
    # [B S E]
    k_tensor = k_tensor.squeeze(dim=-2).reshape(S, B, H*D).transpose(1, 0)
    # [B S E]
    v_tensor = v_tensor.squeeze(dim=-2).reshape(S, B, H*D).transpose(1, 0)

    # [B S E]
    # v_tensor = v_tensor.squeeze(dim=-2).reshape(S, B, -1).permute(1, 0, 2)
    # [B S S]
    # Project queries, keys, and values
    # Scaled Dot-Product Attention

    # k_s_t [64 8 64 10]
    # q_s [64 8 10 64]
    scores = torch.matmul(q_tensor, k_tensor.transpose(-2, -1))/(D ** 0.5)
    attn = F.softmax(scores, dim=-1)

    # Weighted Sum
    # attn [64 8 10 10]
    # v_s [64 8 10 64]
    # weighted_sum[64 8 10 64]
    weighted_sum = torch.matmul(attn, v_tensor).transpose(1, 0)
    print(weighted_sum)

def set_tensor_name(tensor, prefix, name):
    tensor.name = prefix + name

def set_output_name(layer, prefix, name, out_idx=0):
    set_tensor_name(layer.get_output(out_idx), prefix, name)

def build_engine(batch_sizes, workspace_size):
    explicit_batch_flag = 1 << int(
        trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
    with trt.Builder(TRT_LOGGER) as builder, builder.create_network(explicit_batch_flag) as network, builder.create_builder_config() as builder_config:
        builder_config.max_workspace_size = workspace_size * (1024 * 1024)
        builder_config.set_flag(trt.BuilderFlag.FP16)
        # builder_config.clear_flag(trt.BuilderFlag.TF32)
        builder_config.profiling_verbosity = trt.ProfilingVerbosity.DETAILED
        builder_config.set_flag(trt.BuilderFlag.STRICT_TYPES)
        input_tensor_fp32 = network.add_input(
            name="input", dtype=trt.float32, shape=(S, B, E, 1, 1))
        # 或者使用 NumPy 数组初始化权重矩阵
        # W_data = np.random.randn(input_size * num_neurons).astype(np.float32)

        # output_tensor_fp16 = add_fp32_to_fp16_layer(network, input_tensor_fp32)

        W_aout_trt = trt.Weights(np.ascontiguousarray(W_aout))
        attention_out_fc = network.add_fully_connected(
            input_tensor_fp32, 3 * E, W_aout_trt, B_aout)

        attention_out_fc.precision = trt.float16
        fc_output = attention_out_fc.get_output(0)
        set_output_name(attention_out_fc, "fc_", "output")

        # QKV2CTX
        pf_type = trt.PluginField("type_id", np.array(
            [1], np.int32), trt.PluginFieldType.INT32)
        pf_hidden_size = trt.PluginField("hidden_size", np.array(
            [D], np.int32), trt.PluginFieldType.INT32)
        pf_num_heads = trt.PluginField("num_heads", np.array(
            [H], np.int32), trt.PluginFieldType.INT32)
        pf_has_mask = trt.PluginField("has_mask", np.array(
            [0], np.int32), trt.PluginFieldType.INT32)
        pfc = trt.PluginFieldCollection(
            [pf_hidden_size, pf_num_heads, pf_has_mask, pf_type])
        qkv2ctx_plug = qkv2_plg_creator.create_plugin("qkv2ctx", pfc)
        qkv_in = [fc_output]
        qkv2ctx = network.add_plugin_v2(qkv_in, qkv2ctx_plug)
        set_output_name(qkv2ctx, "qkv2ctx_", "output")
        qkv2ctx.precision = trt.float16
        network.mark_output(fc_output)
        network.mark_output(qkv2ctx.get_output(0))
        build_start_time = time.time()
        engine = builder.build_engine(network, builder_config)
        build_time_elapsed = (time.time() - build_start_time)
        TRT_LOGGER.log(TRT_LOGGER.INFO,
                       "build engine in {:.3f} Sec".format(build_time_elapsed))
        TRT_LOGGER.log(TRT_LOGGER.VERBOSE, "Serializing Engine...")
        serialized_engine = engine.serialize()
        TRT_LOGGER.log(TRT_LOGGER.INFO,
                       "Saving Engine to {:}".format(trt_output_model))
        with open(trt_output_model, "wb") as fout:
            fout.write(serialized_engine)
        TRT_LOGGER.log(TRT_LOGGER.INFO, "Done.")
        return engine

engine_trt = build_engine(1, 4096)

trt_outputs = []
trt_outputs_info = []
trt_inputs_info = []
trt_outputs_tensort = {}
dataType = [np.float32, np.float16, np.int8, np.int32]
with engine_trt as engine, engine.create_execution_context() as context:
    inputs, outputs, bindings, stream = common.allocate_buffers(engine)
    for index in range(len(bindings)):
        tensor_name = engine.get_binding_name(index)
        tensor_index = engine.get_binding_index(tensor_name)
        tensor_shape = engine.get_binding_shape(tensor_index)
        tensor_data_type = engine.get_binding_dtype(tensor_index)
        if (engine.binding_is_input(tensor_index)):
            trt_inputs_info.append(
                (tensor_name, tensor_index, tensor_shape, tensor_data_type.value))
        else:
            trt_outputs_info.append(
                (tensor_name, tensor_index, tensor_shape, tensor_data_type.value))

    for trt_input_info in trt_inputs_info:
        print(trt_input_info[0])
        shape = tuple([dim for dim in trt_input_info[2]])
        input_data = input_tensor.unsqueeze(-1).unsqueeze(-1).detach().numpy()
        inputs[trt_input_info[1]].host = input_data
        trt_outputs = common.do_inference_v2(
            context, bindings=bindings, inputs=inputs, outputs=outputs, stream=stream)
        for index in range(len(trt_outputs)):
            shape = tuple([dim for dim in trt_outputs_info[index][2]])
            trt_outputs_tensort[trt_outputs_info[index]
                                [0]] = trt_outputs[index].reshape(shape)

        print("FC max error:", np.max(
            trt_outputs_tensort[trt_outputs_info[0][0]]-qkv_tensor.detach().numpy().reshape([S, B, 3*E, 1, 1])))

        print("trt:", trt_outputs_tensort[trt_outputs_info[1][0]])

        print("torch:", weighted_sum.detach(
        ).numpy().reshape([S, B, E, 1, 1]))

        '''
        print("attention max error:", trt_outputs_tensort[trt_outputs_info[1][0]] -
              attention_output.detach().numpy().reshape([S, B, E, 1, 1]))
        '''

**Comment by zerollzeng** [comment]
@ttyio ^ ^

**Comment by jacker-whh** [comment]
multiHeadFlashAttentionPlugin Why only 8.0, 8.6 and 8.9 hash power, not 8.7?

**Comment by jacker-whh** [comment]
multiHeadFlashAttentionPlugin Why only 8.0, 8.6 and 8.9 hash power, not 8.7?
